### PR TITLE
[ITensors] Consistent precision type of `apply` and `contract`

### PR DESCRIPTION
**Pull request by hjkqubit**
# Description

Previously, if one contracts two `MPO`s with single-precision data type (i.e. Float32) on a 64-bit machine, then the resulting `MPO` will be converted to have Float64 data type due to line 778 of `src/mps/mpo.jl`: `R = ITensor(1)`. 

Changing this line to `R = ITensor(true)` fixes this bug.

<details><summary>Minimal demonstration of previous behavior</summary><p>

```julia
using ITensors
using Random
Random.seed!(1234)

let 
    L = 4
    sites = siteinds("S=1/2",L)
    x = randomMPS(sites,linkdims=1)
    y = randomMPS(sites,linkdims=1)
    A = outer(x',y)

    A = convert_leaf_eltype(Float32, A)
    @show eltype(ITensors.data(apply(A,A))[1])
end
```

returned

```julia
eltype((ITensors.data(A))[1]) = Float32
eltype((ITensors.data(apply(A, A)))[1]) = Float64
```
</p></details>

<details><summary>Minimal demonstration of new behavior</summary><p>

Now, it returns
```julia
eltype((ITensors.data(A))[1]) = Float32
eltype((ITensors.data(apply(A, A)))[1]) = Float32
```
</p></details>

# How Has This Been Tested?

See above.

# Checklist:

- [x] My code follows the style guidelines of this project. Please run `using JuliaFormatter; format(".")` in the base directory of the repository (`~/.julia/dev/ITensors`) to format your code according to our style guidelines.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that verify the behavior of the changes I made.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no new warnings.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

**Comment by mtfishman** [comment]
Thanks! You should be able to use `ITensors.scalartype(A)` to get the element type of the ITensors of an MPS/MPO. Could you add a test based on the code in your first post that uses `ITensors.scalartype`?

**Comment by hjkqubit** [comment]
Yes, I've added the tests!

File: test/mpo.jl [hunk 1]
```diff
@@ -767,6 +767,26 @@ end
...
+
+  @testset "consistent precision type of apply for complex" for T in (ComplexF32, ComplexF64)
+    sites = siteinds("S=1/2", 4)
+    A = 1.0*im*randomMPO(sites)
+    B = 1.0*im*randomMPO(sites)
+
+    A = convert_leaf_eltype(T, A)
+    B = convert_leaf_eltype(T, B)
+    @test ITensors.scalartype(apply(A, B)) == T
+  end
```
Comment (mtfishman) [inline test/mpo.jl:789]: Can we combine these into a single loop? For example you could do:
```julia
A = convert_leaf_eltype(T, randn(T) * randomMPO(sites))
B = convert_leaf_eltype(T, randn(T) * randomMPO(sites))
```

**Comment by mtfishman** [comment]
Thanks!

**Pull request merged**